### ISSUE TITLE
5' adapter not trimmed if read ends early

### ISSUE THREAD

**Issue by marcelm** [marcelm/cutadapt#565]
Given a sequence with 5' adapter, eg **ALONGADAPTOR**sequence, if sequence is low quality in the end, or has polyG, cutadapt will trim this sequence into  **ALONGADAPTOR**seq (*1st case*) or **ALONGADAP** (*2nd case*). Then the `-g` argument and remove the adapter in the 1st case, but not in the 2nd case. And will cause adaptor contamination in the filtered reads.

_Originally posted by @yech1990 in https://github.com/marcelm/cutadapt/issues/550#issuecomment-921100958_

**Comment by marcelm** [comment]
Thanks for reporting, this is indeed a problem that should be fixed.

Some notes to myself and/or those interested in the details.

- This needs to be fixed for three different cases: For regular 5' adapters, for anchored 5' adapters and for anchored 5' adapters in combination with `--no-indels`, which uses a different algorithm.
- When aligning regular 5' adapters, partial matches are only allowed at the 5' end, not at the 3' end. So an adapter `ADAPTER` will be found in `PTERSEQUENCE`, but not in `SEQUENCEADAPT`. This is necessary because otherwise short, random matches at the 3' end would lead to the entire read being trimmed. It is also symmetrical to how 3' adapters are aligned. (For those, partial occurrences at the 5' end are not allowed.)
- Not allowing partial matches at the 3' end is why the adapters are not found in the examples above.
- We therefore need to change the alignment algorithm such that it allows partial matches in those cases where the adapter is longer than the read.
- Idea: For an adapter of length *m* and a read of length *n*, we look in the right-hand column of the alignment matrix not only at row index m, as currently, but also from *n* to *m* if *n* < *m*.

One test case that should work is to find a regular 5' adapter `ADAPTER`, allowing 1 error (ignoring error rate), within `TADAPT`. The alignment could look like this:
```
 ADAPTER
TADAP-
```

**Comment by yech1990** [comment]
Thank you @marcelm

**Comment by yech1990** [comment]
I think I come across a similar problem when dealing with 3' adapter. 

For ssequences like these,

```
sequenceALONGADAPTORGGGGGGGGG
ALONGADAPTORGGGGGGGGGGGGGGGGG
LONGADAPTORGGGGGGGGGGGGGGGGGG
```

when running cutadapt with `-a ALONGADAPTOR`,  the output wiil be

 ```
sequence

LONGADAPTORGGGGGGGGGGGGGGGGGGGGGG
```

I would like to know if it is possible to trim the 3rd sequence ad the 2nd one, and generate a null sequence.

**Comment by marcelm** [comment]
Yes, you can use the `anywhere` adapter-trimming parameter. Jus twrite `-a 'ALONGADAPTOR;anywhere'`. This works:
```
$ echo -e '>r\nADAPTORGGGGGGGGGGGGGGGGGG' | cutadapt -N --quiet -a 'ALONGADAPTOR;anywhere' -
>r

```
(The third example, as you gave it, will actually work because the error rate is by default 0.1, which will allow to delete the initial `L`.)

**Comment by yech1990** [comment]
thank you very much. I found this argument in the document now. 😂

**Comment by y9c** [comment]
HI @marcelm 

Is this issue solved?

**Comment by marcelm** [comment]
No, sorry, I haven’t prioritized it.
